File: leetcode/0030.go
package leetcode

func findSubstring(s string, words []string) []int {
	slen, wnum := len(s), len(words)
	if slen <= 0 || wnum <= 0 {
		return []int{}
	}
	wlen := len(words[0])
	if wlen*wnum > slen {
		return []int{}
	}

	wmap, ans := make(map[string]int), make([]int, 0)
	for _, word := range words {
		wmap[word]++
	}

	for i := 0; i < wlen; i++ {
		smap, count := make(map[string]int), 0
		for l, r := i, i; l <= slen-wlen*wnum && r <= slen-wlen; r += wlen {
			temp := s[r : r+wlen]
			if num, ok := wmap[temp]; ok {
				smap[temp]++
				count++
				for smap[temp] > num {
					smap[s[l:l+wlen]]--
					l += wlen
					count--
				}
				if count == wnum {
					ans = append(ans, l)
					smap[s[l:l+wlen]]--
					count--
					l += wlen
				}
			} else {
				l = r + wlen
				count = 0
				smap = make(map[string]int)
			}
		}
	}
	return ans
}

File: leetcode/0030.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string> &words) {
        int slen = s.size(), wnum = words.size();
        vector<int> ans;
        if (!slen || !wnum) return ans;
        int wlen = words[0].size();
        if (slen < wnum * wlen) return ans;

        unordered_map<string, int> wmap, smap;
        for (auto i : words) wmap[i]++;

        for (int i = 0; i < wlen; i++) {
            smap.clear();
            int count = 0;
            for (int l = i, r = i; l <= slen - wlen * wnum && r <= slen - wlen;
                 r += wlen) {
                string temp = s.substr(r, wlen);
                if (wmap[temp]) {
                    smap[temp]++;
                    count++;
                    while (wmap[temp] < smap[temp]) {
                        smap[s.substr(l, wlen)]--;
                        l += wlen;
                        count--;
                    }
                    if (count == wnum) {
                        ans.push_back(l);
                        smap[s.substr(l, wlen)]--;
                        l += wlen;
                        count--;
                    }
                } else {
                    count = 0;
                    smap.clear();
                    l = r + wlen;
                }
            }
        }
        return ans;
    }
};